Add code to support the new notification system starting with iOS 10

fridtjof opened the pull request:
WARNING: I couldn't test any of this, because I can't build this app for anything except the simulator (I don't have a paid dev membership - necessary for some capabilities the app has, and also I don't have (for obvious reasons) the entitlements that would allow me to easily receive push notifications from the "official" New Vector push server, which is needed to test any notification related behaviour).
Therefore, I'm kindly asking for someone from the team to test my changes on a device with iOS 10 or higher. Feedback is much appreciated :)

This is the first step in supporting newer notifications features. By using the UserNotifications framework, we will be able to implement new notification features, like titles, grouping and more (see https://github.com/vector-im/riot-ios/issues/2208)

Signed-off-by: Fridtjof Mund 2780577+fridtjof@users.noreply.github.com

SBiOSoftWhare commented on Riot/AppDelegate.m:
```diff
@@ -1076,23 +1078,56 @@ - (void)registerUserNotificationSettings
+                                          MXKAccountManager* accountManager = [MXKAccountManager sharedManager];
+                                          [accountManager setPushDeviceToken:nil withPushOptions:nil];
+                                      }
+                                  }];
```
We should also listen to notification settings changes that can occur while the application is running:
```
[center getNotificationSettingsWithCompletionHandler:^(UNNotificationSettings * _Nonnull settings) {
    if (settings.authorizationStatus == UNAuthorizationStatusAuthorized)
    {
        [self registerForRemoteNotificationsWithCompletion:nil];
    }
    else
    {
        // Clear existing token
        MXKAccountManager* accountManager = [MXKAccountManager sharedManager];
        [accountManager setPushDeviceToken:nil withPushOptions:nil];
    }
}];
```

fridtjof replied:
When should this be triggered? Periodically? Or each time the app goes into foreground? Or on every background sync?

SBiOSoftWhare replied:
Sorry you can ignore this comment I will mark it as resolved. This is something else we have to manage.
In fact notification settings could be changed at any time we have to find a way to check those changes periodically as there is no dedicated delegate method for this at this time. For example check for notification settings with in `applicationDidBecomeActive:`.

SBiOSoftWhare commented on Riot/AppDelegate.m:
```diff
@@ -1120,6 +1156,70 @@ - (void)application:(UIApplication *)application didRegisterUserNotificationSett
+    UNNotification *notification = response.notification;
+    UNNotificationContent *content = notification.request.content;
+    if ([[response actionIdentifier] isEqualToString:@"inline-reply"]) {
+        UNTextInputNotificationResponse *textResponse = (UNTextInputNotificationResponse *) response;
```
Note:
Could be interesting to factorize inline reply code between the two APIs with a method signature like this for example:
`- (void)handleNotificationInlineReplyForRoomId:(NSString*)roomId withResponseText:(NSString*)responseText success:(void(^)(NSString *eventId))success failure:(void(^)(NSError *error))failure`.

SBiOSoftWhare commented on Riot/AppDelegate.m:
```diff
@@ -1120,6 +1156,70 @@ - (void)application:(UIApplication *)application didRegisterUserNotificationSett
+    }
+    else
+    {
+        NSLog(@"[AppDelegate][Push] didReceiveNotificationResponse: unhandled identifier %@", [response actionIdentifier]);
```
`userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler:` is also called when user tap on the notification. As fallback case here he should navigate to the notification room id. Like done in `application:didReceiveLocalNotification:`.

fridtjof replied:
https://github.com/vector-im/riot-ios/pull/2207/commits/1c90e0cb4427b76c19457d8df2c3341bc01a031e

SBiOSoftWhare commented on Riot/AppDelegate.m:
```diff
@@ -1120,6 +1156,70 @@ - (void)application:(UIApplication *)application didRegisterUserNotificationSett
+
+                    completionHandler();
+                }];
+            }
```
You could add a `return;` at the end of this else like in `application:handleActionWithIdentifier: forLocalNotification:withResponseInfo:completionHandler:` to be sure `completionHandler();` is called only once.

fridtjof replied:
https://github.com/vector-im/riot-ios/commit/2f1b6153a8190ea4024db8cdb243b8d4338df5ea

SBiOSoftWhare commented on Riot/AppDelegate.m:
```diff
@@ -1352,7 +1500,15 @@ - (void)handleLocalNotificationsForAccount:(MXKAccount*)account
-        if ([self displayedLocalNotificationForEvent:eventId andUser:account.mxCredentials.userId type:nil])
+        if (@available(iOS 10, *))
+        {
+            if ([self displayedNotificationRequestForEvent:eventId andUser:account.mxCredentials.userId type:nil])
```
We are not obliged to check for displayed notifications here, as `UNNotificationRequest` have a unique identifier (evendId). The content for a `UNNotificationRequest` with the same identifier will be overwritten.

fridtjof replied:
https://github.com/vector-im/riot-ios/pull/2207/commits/93ce20e9cc379762acb626bdecfec62d7b2200f5

SBiOSoftWhare commented on Riot/AppDelegate.m:
```diff
@@ -1421,41 +1577,88 @@ - (void)handleLocalNotificationsForAccount:(MXKAccount*)account
-                                                   @"event_id": event.eventId,
-                                                   @"user_id": account.mxCredentials.userId
-                                                   };
+                    if (@available(iOS 10, *))
```
Note:
We could factorize for the 2 APIs like this:
```
NSDictionary *notificationUserInfo = @{
        @"type": @"full",
        @"room_id": event.roomId,
        @"event_id": event.eventId,
        @"user_id": account.mxCredentials.userId
};

BOOL isNotificationContentShown = !event.isEncrypted || RiotSettings.shared.showDecryptedContentInNotifications;

NSString *categoryIdentifier;

if ((event.eventType == MXEventTypeRoomMessage || event.eventType == MXEventTypeRoomEncrypted) && isNotificationContentShown)
{
    categoryIdentifier = @"QUICK_REPLY";
}

NSString *soundName;

// Set sound name based on the value provided in action of MXPushRule
for (MXPushRuleAction *action in rule.actions)
{
    if (action.actionType == MXPushRuleActionTypeSetTweak && [action.parameters[@"set_tweak"] isEqualToString:@"sound"])
    {
        NSString *soundName = action.parameters[@"value"];
        if ([soundName isEqualToString:@"default"])
            soundName = @"message.mp3";     
    }
}
```

fridtjof replied:
https://github.com/vector-im/riot-ios/pull/2207/commits/589829d0765f3a392bce5ed3d962a0489f441765

SBiOSoftWhare commented on Riot/AppDelegate.m:
```diff
@@ -1622,7 +1825,15 @@ - (void)handleLimitedLocalNotifications:(MXSession*)mxSession events:(NSArray<NS
-        if ([self displayedLocalNotificationForEvent:eventId andUser:userId type:nil])
+        if (@available(iOS 10, *))
+        {
+            if ([self displayedNotificationRequestForEvent:eventId andUser:userId type:nil])
```
Like said above:
We are not obliged to check for displayed notifications here, as UNNotificationRequest have a unique identifier (evendId). The content for a UNNotificationRequest with the same identifier will be overwritten.

fridtjof replied:
https://github.com/vector-im/riot-ios/pull/2207/commits/04e9bf9a6b8e768f1a957ef7dff592572c98748c

SBiOSoftWhare commented on Riot/AppDelegate.m:
```diff
@@ -1692,6 +1917,41 @@ - (nullable NSString *)limitedNotificationBodyForEvent:(NSString *)eventId inMat
+ @return the local notification request if any.
+ */
+// iOS 10+ only!
+- (UNNotificationRequest *)displayedNotificationRequestForEvent:(NSString *)eventId andUser:(NSString *)userId type:(NSString*)type
```
This method will always return nil. In fact `foundRequest` value will be returned at the end of the function before 
`getDeliveredNotificationsWithCompletionHandler:` completes, because this method is asynchronous.
But as mentioned above we could remove this method and rely on `UNNotificationRequest` identifier to avoid notifications duplication.

fridtjof replied:
https://github.com/vector-im/riot-ios/pull/2207/commits/90188cdcf36e82116580f0ea8fe1bd3367a0f089

SBiOSoftWhare commented on Riot/AppDelegate.m:
```diff
@@ -1176,6 +1276,54 @@ - (void)application:(UIApplication *)application handleActionWithIdentifier:(NSS
     completionHandler();
 }
 
+// iOS 10+, see application:didReceiveLocalNotification:
```
In fact this method is not equivalent to `application:didReceiveLocalNotification`, when the user tap on a local notification with `UILocalNotification` it calls at the end `application:didReceiveLocalNotification`. But with `UserNotifications` it calls `-[UNUserNotificationCenterDelegate userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler:]`

fridtjof replied:
https://github.com/vector-im/riot-ios/pull/2207/commits/1c90e0cb4427b76c19457d8df2c3341bc01a031e

fridtjof replied:
https://github.com/vector-im/riot-ios/pull/2207/commits/7ae3070e0ae39b0ff50c33f11ac7597e8e3ad50c

SBiOSoftWhare commented on Riot/AppDelegate.m:
```diff
@@ -1176,6 +1276,54 @@ - (void)application:(UIApplication *)application handleActionWithIdentifier:(NSS
+- (void)userNotificationCenter:(UNUserNotificationCenter *)center willPresentNotification:(UNNotification *)notification withCompletionHandler:(void (^)(UNNotificationPresentationOptions options))completionHandler
+{
+    NSLog(@"[AppDelegate][Push] willPresentNotification: applicationState: %@", @([UIApplication sharedApplication].applicationState));
+
```
Missing call to completion handler here. As we currently do nothing when we receive a notification when the application is in foreground we should use `completionHandler(UNNotificationPresentationOptionNone);`. Indeed this delegate method is called when a notification is delivered to a foreground app. 

We could remove all code related to room navigation and just keep this line `completionHandler(UNNotificationPresentationOptionNone);`. There is no reason to navigate to a room each time we receive a notification in foreground.
Instead we should handle navigation to a room when user tap the notification in `userNotificationCenter: didReceiveNotificationResponse:withCompletionHandler:` as default action.

fridtjof replied:
https://github.com/vector-im/riot-ios/pull/2207/commits/7ae3070e0ae39b0ff50c33f11ac7597e8e3ad50c

fridtjof commented:
I forgot to push one slight refactor (https://github.com/vector-im/riot-ios/commit/8cb6b8600dba52cb7ae52b316592be669684327f) - this duplicates code for now, but is necessary in preparation for using the new notification features imo. I can roll it back and reintroduce it later though, if you (understandably) don't want the code duplication.

fridtjof commented:
I addressed most of the reviews now, as it's time for bed now. On each review, I linked the corresponding commit for easy referencing.
I want to add that I'm relatively new to contributing to open source projects, so I appreciate any feedback on how I could improve how I use git(hub) and if there's anything that I should do differently :)

SBiOSoftWhare commented:
> I addressed most of the reviews, as it's time for bed now. On each review, I linked the corresponding commit for easy referencing.
> I want to add that I'm relatively new to contributing to open source projects, so I appreciate any feedback on how I could improve how I use git(hub) and if there's anything that I should do differently :)

Nothing to add, you did a good job all issues has been addressed :) Thank you for your work, I will approve your PR and you will be mentioned in changelog when it will be merged
